### DISCUSSION TITLE
How to prevent layout shift for mobile and desktop versions of site with SSG

### DISCUSSION THREAD

**Discussion by peteole** [vercel/next.js#41035] [Help]
Hi,
I building a simple personal website with next js. I use a card layout pretty often to show various kinds of content.
On mobile devices there should be only one column/line of cards whereas on desktop I like to have multiple columns.
The cards all have the same width but different heights. Therefore I do not want to have a grid layout with rows of cards each starting at the same heights, but I want to have an even spacing around all the cards. See the demo here:
https://ole-petersen.vercel.app/
and the source code here:
https://github.com/peteole/portfolio

The issue that now arises is that there is a layout shift during hydration. The column count needs to be computed based on the device type (or now screen width). I set the default to 1 to focus on mobile devices. This means that on desktop, first the pre-rendered one-column-layout will be loaded and once the javascript takes over, the one, long column shifts over to multiple shorter columns, which is bad UX.

Now I am wondering: Is there a way to statically generate a desktop and a mobile site and serve them based on the user-agent header? If yes how should the config file look like? Or do you know a better way to prevent the layout shift? How do othere people handle different versions of their website on next.js?

I already set a fade-in animation in order to hide the shift, but this does not solve the root problem.
My experiments with rewrites seemed pretty painful...

**Comment by peteole** [comment]
Update: I fixed the issue with a really dirty hack. I now just prerender three versions (one column, two columns, three columns) of the page, but in one file, and use CSS selectors to set `display: none` on all but one of the three versions. Still does not feel like the way this should be handled, so my question remains. To see the layout shift in action have a look at the old version:
https://ole-petersen-dgglgdiih-peteole.vercel.app/

**Reply by wwwdepot** [reply]
In the old version, the columns merged into a single div instead of 2 on a small screen, idk what JS is doing this. This is causing layout shift. 

Use media query for small screen to make columns look like a single column. Avoid using JS to move DOM elements, stick with CSS Flex/Grid if you can.

**Reply by peteole** [reply]
Thanks!
This is kind of what I did now. However for the future I would still like to know if it is possible to pre-render two versions of the website based on the `user-agent` header in case the differences are to big to solve them with CSS?

**Reply by wwwdepot** [reply]
Not sure about that with SSG

**Comment by karandwivedi42** [comment]
Facing the same problem (Next 13 app directory). What is the best practice around this?

**Reply by alex1craig** [reply]
Same issue here.

--- Edit:
I had CSS transitions applied to all for divs that I just wanted to change color on mouse hover. I simply changed "transition: all 0.3s" to "transition: color, background-color 0.3s" for those divs. I think padding and margin were also takings 300ms to get into the right spot on page load which was causing the layout shift. So maybe just take a look at your CSS and make sure you're not doing something similar.
